### PR TITLE
[grid] Add default capabilities to Node config

### PR DESCRIPTION
**Pull request by pujagani**
**Thanks for contributing to Selenium!**
**A PR well described will help maintainers to quickly review and merge it**

Before submitting your PR, please check our [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) guidelines.
Avoid large PRs, help reviewers by making them as simple and short as possible.

### Description

Addressing PR #8761 related to #8745.

### Motivation and Context

Grid 3 had a mechanism for adding default capabilities. Grid 4 uses a different method (AutoService annotation and Toml config) to provide config values. The changes allow the node to have the default capabilities as per Grid 4. If the default capabilities are present then the auto-detection of all drivers is not done. 

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [X] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: java/server/src/org/openqa/selenium/grid/node/config/NodeOptions.java [hunk 1]
```diff
@@ -82,17 +86,63 @@ public Node getNode() {
...
+    infos.forEach(info -> {
+      Capabilities caps = info.getCanonicalCapabilities();
+      builders.stream()
+        .filter(builder -> builder.score(caps) > 0)
+        .forEach(builder -> {
+          for (int i = 0; i < Math.min(info.getMaximumSimultaneousSessions(), maxSessions); i++) {
+            Capabilities canonicalCaps = info.getCanonicalCapabilities();
+            Collection<SessionFactory> sessionFactoryCollection = factoryFactory.apply(info);
+            sessionFactories.putAll(canonicalCaps, sessionFactoryCollection );
+            report(Map.entry(info, sessionFactoryCollection));
```
Comment (link89) [inline java/server/src/org/openqa/selenium/grid/node/config/NodeOptions.java:140]: Error
```
java/server/src/org/openqa/selenium/grid/node/config/NodeOptions.java:140: error: cannot find symbol
            report(Map.entry(info, sessionFactoryCollection));
                      ^
  symbol:   method entry(WebDriverInfo,Collection<SessionFactory>)
  location: interface Map
```


Fixed by `            report(new AbstractMap.SimpleEntry(info, sessionFactoryCollection));`

**Comment by link89** [comment]
It fails when executing `java -jar bazel-bin/java/server/src/org/openqa/selenium/grid/selenium_server_deploy.jar standalone --capabilities '{"browserName":"chrome","goog:chromeOptions":{"binary":"/Applications/Google Chrome Canary.app/Contents/MacOS/Google Chrome Canary"}}'`

```
17:43:54.544 INFO [LoggingOptions.getTracer] - Using OpenTelemetry for tracing
Exception in thread "main" org.openqa.selenium.grid.config.ConfigException: java.lang.reflect.InvocationTargetException
	at org.openqa.selenium.grid.config.MemoizedConfig.getClass(MemoizedConfig.java:115)
	at org.openqa.selenium.grid.node.config.NodeOptions.getNode(NodeOptions.java:80)
	at org.openqa.selenium.grid.commands.Standalone.createHandlers(Standalone.java:195)
	at org.openqa.selenium.grid.TemplateGridServerCommand.asServer(TemplateGridServerCommand.java:41)
	at org.openqa.selenium.grid.commands.Standalone.execute(Standalone.java:206)
	at org.openqa.selenium.grid.TemplateGridCommand.lambda$configure$2(TemplateGridCommand.java:98)
	at org.openqa.selenium.grid.Main.launch(Main.java:154)
	at org.openqa.selenium.grid.Main.go(Main.java:88)
	at org.openqa.selenium.grid.Main.main(Main.java:54)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.openqa.selenium.grid.config.ClassCreation.callCreateMethod(ClassCreation.java:30)
	at org.openqa.selenium.grid.config.MemoizedConfig.lambda$getClass$4(MemoizedConfig.java:100)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1737)
	at org.openqa.selenium.grid.config.MemoizedConfig.getClass(MemoizedConfig.java:95)
```

File: java/server/src/org/openqa/selenium/grid/node/config/NodeOptions.java [hunk 2]
```diff
@@ -82,17 +86,63 @@ public Node getNode() {
...
 
+  private void addDefaultDrivers(
+    int maxSessions, List<String> defaultConfigCapabilities,
+    Function<WebDriverInfo, Collection<SessionFactory>> factoryFactory,
+    ImmutableMultimap.Builder<Capabilities, SessionFactory> sessionFactories) {
+
+    // Convert all string config values into capabilities and store the browser names.
+    List<String> browserNames = new ArrayList<>();
+    for (String caps : defaultConfigCapabilities) {
+      Capabilities stereotype = JSON.toType(caps, Capabilities.class);
```
Comment (link89) [inline java/server/src/org/openqa/selenium/grid/node/config/NodeOptions.java:115]: I think the serializer may leads to the runtime error, try:

```java
      Map<String, Object> obj = new Json().toType(caps, Json.MAP_TYPE);
      return new MutableCapabilities(obj);
```

**Comment by link89** [comment]
Hi @pujagani , I have tried your patch and find that it doesn't work as expected.
Here is the configuration file
```toml
[node]
capabilities = '{"browserName":"chrome","goog:chromeOptions":{"binary":"/Applications/Google Chrome Canary.app/Contents/MacOS/Google Chrome Canary"}}'
```
And it turns out that the default capabilities are not respected when creating a new session.

The motivation of default capabilities is to allow the server-side to decide the default values of those fields when the client-side omits them. So in the end the server-side capabilities should merge with the client-sides before fire the NEW_SESSION request, just as what it does in the origin PR: https://github.com/SeleniumHQ/selenium/blob/d2107c9f7e67cd9aa59981ef452ba10803711cd1/java/server/src/org/openqa/selenium/grid/node/config/DriverServiceSessionFactory.java#L109-L111

File: java/server/src/org/openqa/selenium/grid/node/config/NodeOptions.java (hunk 1 above)
Comment (pujagani) [reply java/server/src/org/openqa/selenium/grid/node/config/NodeOptions.java:140]: The new Grid uses Java 8 and will continue supporting that for the future versions (https://github.com/SeleniumHQ/selenium/blob/0683bb862d419127b432a170600aa3574edc7c6c/.bazelrc#L6). Thank you for pointing out, though Map.Entry is valid for Java 8.

**Comment by pujagani** [comment]
Thank you for providing feedback. With the Grid 4, toml config is used for setting up the Node. Based on options provided in the Node config, the node would be set up. By default, web drivers on the system would be picked up. This can be disabled. However, the main aim of default capabilities in the new Grid is to avoid this auto-detection of drivers if default capabilities are already provided via the config and to respect the default values. As mentioned in https://github.com/SeleniumHQ/selenium/pull/8761#pullrequestreview-513493051. However, currently the changes in this PR are not aligned with the desired approach. More context and understanding is required from my end as well. Selenium team is working towards it.  Converting the PR into a draft until we decide on a design for this.

**Comment by link89** [comment]
Thank you for your reply.

> However, the main aim of default capabilities in the new Grid is to avoid this auto-detection of drivers if default capabilities are already provided via the config and to respect the default values.

IMHO the auto-detection of drivers and the default capabilities should be two different things.

The reason I hope to support default capabilities is we set up a selenium-grid that contains different OS & browsers. The browser would be installed on different paths or require special arguments (customized user-data directory for different OS, for example) which are impossible for the client-side to know. So we need a way to decide those values on the server-side instead of the client-side. To me, the default capabilities machanism is perfect for this purpose.

**Comment by dylanlive** [comment]
I agree with @link89 that it would be nice to be able to support two or more different versions of a browser, and on the node supply the path to that browser version executable.

**Comment by diemol** [comment]
Closed in favor of https://github.com/SeleniumHQ/selenium/commit/d48a9596d868180b8b2da96e3c3118127365d434

**Pull request closed without merging**